python/dolfinx/graph.py

```diff
@@ -42,15 +69,9 @@ def adjacencylist(data: np.ndarray, offsets=None):
 
     Returns:
         An adjacency list.
-
     """
-    if offsets is None:
-        try:
-            return _cpp.graph.AdjacencyList_int32(data)
-        except TypeError:
-            return _cpp.graph.AdjacencyList_int64(data)
-    else:
-        try:
-            return _cpp.graph.AdjacencyList_int32(data, offsets)
-        except TypeError:
-            return _cpp.graph.AdjacencyList_int64(data, offsets)
+    # Switch to np.isdtype(data.dtype, np.int32) once numpy >= 2.0 is enforced
+    is_32bit = data.dtype == np.int32
+    cpp_t = _cpp.graph.AdjacencyList_int32 if is_32bit else _cpp.graph.AdjacencyList_int64
+    cpp_object = cpp_t(data, offsets) if offsets is not None else cpp_t(data)
+    return AdjacencyList(cpp_object)
```

Comment:
So we need a guard for invalid input style (ie np.floating), or do we think the general nanobind error is sufficient? Usually the nanobind type arrays aren’t very expressive when it comes to dtype.

Reply:
Defending against such calls would require quite a bit of additional logic. It doesn't end there if we did that, then we would need for example also checks of the shapes of the arrays.

Given that it can not cause a false positive, i.e. if we call with a `float` array it will not result in a working code path that produces bad results, and type hints being present, that should flag this in a modern IDE or with `mypy`, I think we should stick to the simplicity of not dealing with it.